snapcraft/models/project.py

```diff
@@ -436,13 +437,36 @@ def _validate_autostart_name(cls, name):
 
         return name
 
-    @pydantic.field_validator("bus_name")
+    @pydantic.field_validator(
+        "command", "stop_command", "post_stop_command", "reload_command", "bus_name"
+    )
     @classmethod
-    def _validate_bus_name(cls, name):
-        if not re.match(r"^[A-Za-z0-9/. _#:$-]*$", name):
-            raise ValueError(f"{name!r} is not a valid bus name")
+    def _validate_apps_section_content(cls, command: str, info: pydantic.ValidationInfo) -> str:
+        # Find any invalid characters in the command.
+        # The regex below is derived from snapd's validator code, modified to be the inverse (^).
+        # https://github.com/canonical/snapd/blob/0706e2d0b20ae2bf030863f142b8491b66e80bcb/snap/validate.go#L756
+        bad_chars: set[str] = set()
+        for bad_char in re.finditer(r"[^A-Za-z0-9/. _#:$-]", command):
+            bad_chars.add(bad_char.group(0))
+
+        if bad_chars:
+            # Guaranteed not-none as the pydantic field_validator decorator always populates it.
+            deserialized = cast(str, info.field_name).replace("_", "-")
+            bad_chars_print = ", ".join(sorted(bad_chars))
+            doc_link = (
```

Comment:
A link to docs is a great idea, but I don't think we should do this today.  Instead we should wait until we get the schema reference in RTD and then configure this: https://github.com/canonical/craft-application/blob/dc1928f2e129cfa06df67ad86f2ad430e136dfc1/craft_application/models/base.py#L110

Reply:
Counterpoint: include this link now and update it when we have an RTD link